### DISCUSSION TITLE
How do you set pre-installed kernel?

### DISCUSSION THREAD

**Discussion by CicadaSeventeen** [SpiralLinux/SpiralLinux-project#179] [Q&A]
Hello, I am studying about the files of calamares, but I cannot find kernel packages in your package lists. So how can you make sure that both 5.10 default kernel and 5.18/5.19 backport kernel  are pre-installed?

**Comment by CicadaSeventeen** [comment]
Besides, why you force to use ipv4 and do not install recommends as default?

**Reply by geckolinux** [reply]
> Besides, why you force to use ipv4

Thanks, that was left over from the original [nodisc config](https://github.com/nodiscc/debian-live-config/tree/master/config/includes.chroot/etc/apt/apt.conf.d) that I used as a base. It will be removed in the next release:
https://github.com/SpiralLinux/SpiralLinux-project/issues/185

> and do not install recommends as default?

To reduce bloat. The recommended packages in Debian are just that, recommended, and should not be installed by default.

**Reply by CicadaSeventeen** [reply]
>and do not install recommends as default?
>To reduce bloat. The recommended packages in Debian are just that, recommended, and should not be installed by default.

In fact, in Debian, some of "soft depends" will be tagged as "Recommends", without them program may even not able to work at all, e.g. `podman`  `snapd` `fcitx5`. That is why some people like to set "--install-recommends" as default.

**Reply by geckolinux** [reply]
Right, you're correct. It's not ideal, but it seems best to let users install recommends on their systems as needed. That's why SpiralLinux officially recommends using Synaptic (not the `apt` / `apt-get` commands) for managing packages, where you can simply right-click on a package and optionally install its recommended packages.

**Comment by geckolinux** [comment]
Hi there,

> So how can you make sure that both 5.10 default kernel and 5.18/5.19 backport kernel are pre-installed?

Good question. The kernel packages are defined in the `auto/config` script from Debian `live-build`, which is renamed to `auto/iso` in SpiralLinux for working with the `spiral` wrapper script. There you will see:
`--linux-packages "linux-image linux-headers" \`
I'm not completely sure how `live-build` installs both the 5.10 default kernel as well as the kernel version from Backports, but I'm almost sure that it's from the `/usr/lib/live/build/bootstrap` script. Then, you can see in `/usr/lib/live/build/chroot_linux-image` it contains:
```
# Processing arguments and configuration files
Init_config_data "${@}"

if [ "${LB_LINUX_PACKAGES}" = "none" ]
then
	exit 0
fi

Echo_message "Begin scheduling kernel image installation..."

# Requiring stage file
Require_stagefiles config bootstrap
```
So, I think what happens is that the bootstrap stage installs the default (5.10) kernel images directly from the defined source of Debian Stable main archives before APT exists in the chroot system, so the SpiralLinux `/etc/apt/preferences.d/99backports` preferences are not obeyed yet. Then, later in the build process after bootstrap it installs the rest of the system with APT and it obeys the preference for the `linux-image-amd64` image from Backports and installs that too.

**Reply by CicadaSeventeen** [reply]
Thanks. 
Your explanation does make sense and give me some idea.
Actually `linux-image-amd64` and `linux-headers-amd64` are "empty packages". They use `Depends` to point to the real kernel package with different versions, so if kernel packages are updated then the old kernel will not be removed. In Stable source and Backports source there are two different version of them but with the same package names. So bootstrap install packages from Stable then it will install 5.10 kernel and then Backports repository is added here then even an `apt upgrade`  can update the `linux-image-amd64` / `linux-headers-amd64` to Backports version so 5.19 kernel will be install.
